### ISSUE TITLE
`PYI036` false positive on stringized (quoted) `__exit__` / `__aexit__` annotations

### ISSUE THREAD

**Issue by thormick** [astral-sh/ruff#27755]
## Summary

`PYI036` (`bad-exit-annotation`) inspects the raw annotation expression and never looks through a string literal, so it reports a violation for *every* stringized form of an otherwise-correct annotation. This also includes what it's own message for `TC004` on `types.TracebackType` asks for.

A stringized annotation means the same thing to a type checker as the unquoted one, so all four diagnostics below are false positives.

## Minimal reproduction

`minimal.py`:

```python
import types
from types import TracebackType


class Quoted:
    def __exit__(
        self,
        typ: type[BaseException] | None,
        exc: BaseException | None,
        tb: "TracebackType | None",
    ) -> None: ...

    async def __aexit__(
        self,
        typ: type[BaseException] | None,
        exc: BaseException | None,
        tb: "TracebackType | None",
    ) -> None: ...


class QuotedQualified:
    def __exit__(
        self,
        typ: type[BaseException] | None,
        exc: BaseException | None,
        tb: "types.TracebackType | None",
    ) -> None: ...


class QuotedObject:
    def __exit__(
        self,
        typ: type[BaseException] | None,
        exc: BaseException | None,
        tb: "object",
    ) -> None: ...


class Unquoted:  # control: correctly produces no diagnostic
    def __exit__(
        self,
        typ: type[BaseException] | None,
        exc: BaseException | None,
        tb: TracebackType | None,
    ) -> None: ...
```

```console
$ ruff check --isolated --no-cache --select PYI036 minimal.py
PYI036 The third argument in `__exit__` should be annotated with `object` or `types.TracebackType | None`
  --> minimal.py:10:13
PYI036 The third argument in `__aexit__` should be annotated with `object` or `types.TracebackType | None`
  --> minimal.py:17:13
PYI036 The third argument in `__exit__` should be annotated with `object` or `types.TracebackType | None`
  --> minimal.py:26:13
PYI036 The third argument in `__exit__` should be annotated with `object` or `types.TracebackType | None`
  --> minimal.py:35:13
Found 4 errors.
```

Expected: no diagnostics. The `Unquoted` control is handled correctly, which isolates the string literal as the trigger.

## Reproduction through ruff autofix

The real bummer is that ruff autofix can produce this:

`ruff.toml`:

```toml
target-version = "py312"

[lint]
select = ["PYI", "TC"]

[lint.flake8-type-checking]
quote-annotations = true
```

`chain.py`:

```python
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from types import TracebackType


class Demo:
    def __exit__(
        self,
        typ: type[BaseException] | None,
        exc: BaseException | None,
        tb: TracebackType | None,
    ) -> None: ...
```

```console
$ ruff check --no-cache chain.py
TC004 Quote references to `types.TracebackType`. Import is in a type-checking block.
...
Found 1 error.

$ ruff check --no-cache --fix --unsafe-fixes chain.py
...
Found 2 errors (1 fixed, 1 remaining).

# ruff rewrote line 12 to:  tb: "TracebackType | None",

$ ruff check --no-cache chain.py
PYI036 The third argument in `__exit__` should be annotated with `object` or `types.TracebackType | None`
...
Found 1 error.
```

So `ruff check --fix` turns a `TC004` violation into a `PYI036` violation, and the only remaining escapes are a `noqa`/suppression comment, or adding `from __future__ import annotations` to every affected module.

### Version

0.16.3
